docs/extensions/datacollector.md

```diff
@@ -0,0 +1,167 @@
+            DataCollectionEnvironmentContext environmentContext)
+        {
+            events.SessionStart += this.SessionStarted_Handler;
+            events.TestCaseStart += this.Events_TestCaseStart;
+            logFileName = configurationElement["LogFileName"];
+        }
+    
+    private void SessionStarted_Handler(object sender, SessionStartEventArgs args)
+        {
+            var filename = Path.Combine(AppContext.BaseDirectory, logFileName);
+            File.WriteAllText(filename, "SessionStarted");
+            this.dataCollectionSink.SendFileAsync(this.context.SessionDataCollectionContext, filename, true);
+            this.logger.LogWarning(this.context.SessionDataCollectionContext, "SessionStarted");
+        }
+
+
+    private void Events_TestCaseStart(object sender, TestCaseStartEventArgs e)
+        {
+            this.logger.LogWarning(this.context.SessionDataCollectionContext, "TestCaseStarted " + e.TestCaseName);
+        }
+}
+```
+TestPlatform uniquely identifies each of the DataCollector by `DataCollectorFriendlyName` and `DataCollectorTypeUri`.
+
+Here is a brief description of each of the argument that is passed in constructor.
+
+### Configuration Element
+At times, there is a need to pass configuration settings for initializing data collectors that could vary between test runs. 
+E.g. module names to exclude from code coverage.
+For supporting those scenarios, configuration xml can be passed to DataCollectors using runsettings.
+```xml
+<RunSettings>
+    <DataCollectionRunSettings>
+        <DataCollectors>
+            <DataCollector friendlyName="NewDataCollector">
+                <Configuration>
+                    <LogFileName>DataCollectorLogs.txt</LogFileName>
+                </Configuration>
+            </DataCollector>
+        </DataCollectors>
+    </DataCollectionRunSettings>
+</RunSettings>
+```
+```csharp
+XmlElement logFileElement = configurationElement[LogFileName];
+string logFile = logFileElement != null ? logFileElement.InnerText : string.Empty;
+if (!File.Exists(logFile))
+    {
+            // Create a file to write to.
+            string createText = "Hello and Welcome" + Environment.NewLine;
+            File.WriteAllText(path, createText);
+    }
+```
+
+### DataCollectionEvents
+DataCollectors can choose to subscribe to any of the four events exposed by `DataCollectionEvents` 
+1. TestSessionStart : Raised when test execution session starts.
+2. TestSessionEnd : Raised when test execution session ends.
+3. TestCaseStart : Raised when test case execution starts.
+4. TestCaseEng : Raised when test case execution ends.
+
+```csharp
+events.SessionStart += this.SessionStarted_Handler;
+events.SessionEnd += this.SessionEnded_Handler;
+events.TestCaseStart += this.Events_TestCaseStart;
+events.TestCaseEnd += this.Events_TestCaseEnd;
+```
+```csharp
+private void Events_TestCaseStart(object sender, TestCaseStartEventArgs e)
+{
+}
+```
+### DataCollectionSink
```

Comment:
We should explain what are `Attachments`. When should a data collector use attachments? Also attachments are available to loggers for post processing.